src/aead/aead.rs

```diff
@@ -76,37 +76,24 @@ impl OpeningKey {
     pub fn algorithm(&self) -> &'static Algorithm { self.key.algorithm() }
 }
 
-/// Authenticates and decrypts (&ldquo;opens&rdquo;) data in place.
-///
-/// The input is `in_out[in_prefix_len..]`; i.e. the input is the part of
-/// `in_out` after the prefix. When `open_in_place` returns `Ok(out_len)`, the
-/// decrypted output is `in_out[..out_len]`; i.e. the output has been written
-/// over the top of the prefix and the input. To put it a different way, the
-/// output overwrites the input, shifted by `in_prefix_len` bytes. To have the
-/// output overwrite the input without shifting, pass 0 as `in_prefix_len`.
-/// (The input/output buffer is expressed this way because Rust's type system
-/// does not allow us to have two slices, one mutable and one immutable, that
-/// reference overlapping memory at the same time.)
+/// Authenticates and decrypts (&ldquo;opens&rdquo;) data in place. When
+/// `open_in_place` returns `Ok(plaintext)`, the decrypted output is
+/// `plaintext`, which is a slice of `in_out`.
 ///
 /// C analog: `EVP_AEAD_CTX_open`
 ///
 /// Go analog: [`AEAD.Open`](https://golang.org/pkg/crypto/cipher/#AEAD)
-pub fn open_in_place(key: &OpeningKey, nonce: &[u8], in_prefix_len: usize,
-                     in_out: &mut [u8], ad: &[u8])
-                     -> Result<usize, error::Unspecified> {
+pub fn open_in_place<'a>(key: &OpeningKey, nonce: &[u8], ad: &[u8],
```

Comment:
> Why split? When they received it from the seal_in_place() function, it was already separate, wasn't it?

In TLS and similar protocols, you get the length of `ciphertext+tag` as an input, and then you have to split them.

I see that JWE (JOSE in general, it seems) and probably XML security stuff doesn't work that way. See for example, https://tools.ietf.org/html/rfc7516#section-3.1 and https://tools.ietf.org/html/rfc7516#section-3.2.

> But other users could find simpler ways to keep ciphertext and tag separate. Like a struct Message { sender, time, ..., ciphertext: Vec<u8>, tag: Vec<u8>, ... }, which they can even serialize to JSON or XML and read with their AEAD implementation in Javascript or elsewhere. For these users, it would be weird if they had to concatenate the two.

That's what I'd like to discourage, because such applications have to keep two items (ciphertext and tag) in sync, instead of one item (ciphertext and tag). JOSE and WS-Security are extraordinarily terrible examples to follow in this respect.

So I guess we need to decide about whether we want to make it simpler to implement bad protocols (JOSE and WS-Security), or make it simpler to design protocols correctly.